### PR TITLE
Update unsloth for torch.cuda.amp deprecation

### LINKED ISSUE #2015: Unsloth decorators are deprecated
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

When importing unsloth.py with a recent version of PyTorch, the code should load without warnings or deprecation messages.

### Current behaviour

Whenever `src/axolotl/utils/gradient_checkpointing/unsloth.py` is imported when a recent version of PyTorch is installed, this deprecation warning appears:

```
10.10.10.2: /home/alyssa/lm_fun/qlora-pipe/axolotl/src/axolotl/utils/gradient_checkpointing/unsloth.py:28: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
10.10.10.2:   @torch.cuda.amp.custom_fwd
10.10.10.2: /home/alyssa/lm_fun/qlora-pipe/axolotl/src/axolotl/utils/gradient_checkpointing/unsloth.py:39: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
10.10.10.2:   @torch.cuda.amp.custom_bwd
```

### Steps to reproduce

```
git clone https://github.com/axolotl-ai-cloud/axolotl.git
cd axolotl/
cd src/axolotl/utils/gradient_checkpointing/
python
import unsloth
```

(if a recent PyTorch is installed the deprecation message will appear)

### Config yaml

```yaml
N/A
```

### Possible solution

Unsloth has an upstream fix here:

https://github.com/unslothai/unsloth/blob/main/unsloth/kernels/utils.py

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12

### axolotl branch-commit

main

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

### PR DESCRIPTION
**Pull request by bursteratom**
# Description

torch.cuda.amp is deprecated in newer version of Pytorch. This PR updates unsloth.py to reflect that. Fixes [issue#2015](https://github.com/axolotl-ai-cloud/axolotl/issues/2015)

## Motivation and Context

`axolotl/src/axolotl/utils/gradient_checkpointing/unsloth.py:28: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
axolotl/src/axolotl/utils/gradient_checkpointing/unsloth.py:39: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
10.10.10.2:   @torch.cuda.amp.custom_bwd`

## How has this been tested?

manually imported unsloth and now there are no more warning of deprecation

### CODE REVIEW

**Comment by winglian** [comment]
it seems that these changes causes some subset of the test matrix to take significantly longer to complete.

File: src/axolotl/utils/gradient_checkpointing/unsloth.py [hunk 1]
```diff
@@ -13,8 +13,19 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+from importlib.metadata import version
+
 import torch
 
+torch_version = version("torch")
+
+if torch_version < "2.4.0":
```
Comment (winglian) [inline src/axolotl/utils/gradient_checkpointing/unsloth.py]: `torch.version` is brittle, let's use some other version parsing check

Comment (winglian) [reply src/axolotl/utils/gradient_checkpointing/unsloth.py]: similar to the change needed for the ADOPT optimizer PR, see https://github.com/huggingface/transformers/blob/main/src/transformers/pytorch_utils.py#L20-L39

Comment (bursteratom) [reply src/axolotl/utils/gradient_checkpointing/unsloth.py]: fixed

**Comment by bursteratom** [comment]
> it seems that these changes causes some subset of the test matrix to take significantly longer to complete.
@winglian 
I removed `partial` wrapper and used torch.amp 's internal `torch.amp.custom_fwd(device_type="cuda")` , let's see if this helps with the speed

**Review by winglian — APPROVED** [review]
